vllm/compilation/qk_norm_rope_fusion.py

```diff
@@ -0,0 +1,278 @@
+                FUSED_QK_ROPE_OP,
+                qkv=qkv,
+                num_heads_q=self.num_heads,
+                num_heads_k=self.num_kv_heads,
+                num_heads_v=self.num_kv_heads,
+                head_dim=self.head_dim,
+                eps=self.eps,
+                q_weight=q_weight,
+                k_weight=k_weight,
+                cos_sin_cache=cos_sin_cache,
+                is_neox=self.is_neox,
+                position_ids=pos_flat,
+            )
+            result_qkv = result[1]
+
+            # Split back to q,k,v and return
+            split_tuple = SPLIT_SIZES_OP(
+                result_qkv, [self.q_size, self.kv_size, self.kv_size], -1
+            )
+            return (
+                operator.getitem(split_tuple, 0),
+                operator.getitem(split_tuple, 1),
+                operator.getitem(split_tuple, 2),
+            )
+
+        # Sample inputs to help pattern tracing
+        T = 5
+        qkv = empty_bf16(T, self.q_size + 2 * self.kv_size)
+        positions = empty_i64(T)
+        q_weight = empty_bf16(1, self.head_dim)
+        k_weight = empty_bf16(1, self.head_dim)
+        cos_sin_cache = empty_bf16(4096, self.head_dim)
+        inputs = [
+            qkv,
+            positions,
+            q_weight,
+            k_weight,
+            cos_sin_cache,
+        ]
+
+        # # Register variants across rope ops and with/without contiguous()
+        # # Ensure view ops are canonicalized to reshape in the traced pattern
+        # def wrap_trace_fn(process_fx, trace_fn):
+        #     def wrapped(*args, **kwargs):
+        #         return process_fx(trace_fn(*args, **kwargs))
+
+        #     return wrapped
+
+        # def fx_view_to_reshape(gm: torch.fx.GraphModule):
+        #     from torch._inductor.fx_passes.post_grad import view_to_reshape
+
+        #     view_to_reshape(gm)
+        #     return gm
+
+        pm.register_replacement(
+            pattern,
+            replacement,
+            inputs,
+            pm.fwd_only,
+            pm_pass,
+        )
+
+
+class QKNormRoPEFusionPass(VllmPatternMatcherPass):
+    """Fuse Q/K RMSNorm + RoPE into fused_qk_norm_rope when the custom op exists."""
+
+    @enable_fake_mode
+    def __init__(self, config: VllmConfig):
+        super().__init__(config)
+        self.patterns: PatternMatcherPass = PatternMatcherPass(
+            pass_name="qk_norm_rope_fusion_pass"
+        )
+
+        if not current_platform.is_cuda_alike():
+            logger.debug("QK Norm+RoPE fusion not enabled: unsupported platform")
+            return
+
+        # Register a pattern per attention layer, as sizes differ by shard
+        attn_layers = get_layers_from_vllm_config(config, Attention)
+        if len(attn_layers) == 0:
+            logger.warning(
+                "QK Norm+RoPE fusion enabled, but no Attention layers were discovered."
+            )
+            return
+        layer_name, layer = next(iter(attn_layers.items()))
```

Comment:
This will only register the pattern using one layer, is that intended? Are we sure this will always pick the same layer also?